src/lib/InstantSearch.js

```diff
@@ -389,75 +312,69 @@ See: https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-o
    * @return {undefined} This method does not return anything
    */
   dispose() {
-    this.removeWidgets(this.widgets);
+    this.removeWidgets(this.mainIndex.getWidgets());
+    this.mainIndex.dispose();
+
     // You can not start an instance two times, therefore a disposed instance
     // needs to set started as false otherwise this can not be restarted at a
     // later point.
     this.started = false;
 
-    this.derivedHelper.detach();
-    this.derivedHelper = null;
-
     // The helper needs to be reset to perform the next search from a fresh state.
     // If not reset, it would use the state stored before calling `dispose()`.
-    this.helper.removeAllListeners();
+    this.removeAllListeners();
+    this.mainHelper.removeAllListeners();
+    this.mainHelper = null;
     this.helper = null;
   }
 
+  scheduleSearch = defer(() => {
```

Comment:
where's this function called?

Reply:
It is called on [`addWidget`](https://github.com/algolia/instantsearch.js/blob/8881841f28240766dded1a49b3dcb89ef3565327/src/lib/index.ts#L106) & [`removeWidget`](https://github.com/algolia/instantsearch.js/blob/8881841f28240766dded1a49b3dcb89ef3565327/src/lib/index.ts#L144) at the index level once the instance is started. An `index` widget could contain one (or many) index widgets. It means that all of them will trigger a new search. To avoid multiple searches we defer them to trigger only one.